flac/decoder.go

```diff
@@ -0,0 +1,109 @@
+// TODO: add seek support; only enabled in the dev branch of mewkiz/flac.
+
+package flac
+
+import (
+	"io"
+
+	"github.com/mewkiz/flac"
+	"github.com/mewkiz/flac/frame"
+	"zikichombo.org/sound"
+	"zikichombo.org/sound/freq"
+)
+
+// Decoder encapsulates state for decoding and seeking a FLAC audio stream.
+type Decoder struct {
+	stream *flac.Stream
+	frame  *frame.Frame // current frame.
+	i      int          // index of current sample in subframe(s).
+}
+
+// NewDecoder creates a decoder from a FLAC audio stream (seekable, readable).
+func NewDecoder(r io.Reader) (*Decoder, error) {
+	stream, err := flac.New(r)
+	if err != nil {
+		return nil, err
+	}
+	d := &Decoder{
+		stream: stream,
+	}
+	return d, nil
+}
+
+func (d *Decoder) Receive(dst []float64) (int, error) {
+	if len(dst)%d.Channels() != 0 {
+		return 0, sound.ErrChannelAlignment
+	}
+	n := 0                       // number of frames (samples per channel) buffered.
+	m := len(dst) / d.Channels() // number of frames (samples per channel) to buffer.
+	bps := int(d.stream.Info.BitsPerSample)
+	for n < m {
+		if d.frame == nil {
+			frame, err := d.stream.ParseNext()
+			if err != nil {
+				return n, err
+			}
+			d.frame = frame
+			d.i = 0
+		}
+		samplesLeft := len(d.frame.Subframes[0].Samples[d.i:])
+		j := m
+		if j > samplesLeft {
+			j = samplesLeft
+		}
+		if n+j > m {
+			j = m - n
+		}
+		if j == 0 {
+			panic("cannot retrieve zero samples")
+		}
+		for c := 0; c < d.Channels(); c++ {
+			toFloats(dst[c*m+n:c*m+n+j], d.frame.Subframes[c].Samples[d.i:d.i+j], bps)
+		}
+		d.i += j
+		n += j
+		if len(d.frame.Subframes[0].Samples[d.i:]) == 0 {
+			d.frame = nil
+			d.i = 0
+		}
+	}
+	if n < m {
+		// TODO: handle n < m. Move frames in dst to their correct position
+		//
+		//    dst[c*n:(c+1)*n]
+		panic("flac.Decoder.Receive: reached end of stream with n < m; support for this case is not yet implemented;\n\nThe implementation should move frames (in ziki terminology) in dst to their correct position, i.e.\n\n   dst[c*n:(c+1)*n]")
+	}
+	return n, nil
+}
+
+func (d *Decoder) Channels() int {
+	return int(d.stream.Info.NChannels)
+}
+
+func (d *Decoder) SampleRate() freq.T {
+	return freq.T(d.stream.Info.SampleRate) * freq.Hertz
+}
+
+func (d *Decoder) Close() error {
+	return d.stream.Close()
+}
+
+// ### [ Helper functions ] ####################################################
+
+// Copied from sound/sample/fix.go and adjusted from int64 to int32.
+
+func toFloat(d int32, nBits int) float64 {
+	s := float64(int32(1 << uint(nBits-1)))
+	return float64(d) / s
+}
+
+func toFloats(dst []float64, src []int32, nBits int) []float64 {
```

Comment:
Move functionality into `zc/sound/sample`?

Reply:
SGTM
ToFloat[s]From32?
ToFloat[s]32?

Reply:
I'd put an 's' after Int32 to be consistent with Float+s

Do you think that only Int32 needs name distinguished?  At present, as there is only int64 support, int64 doesn't say "64" or "int64" in the name.  It might be more consistent to name them both with the size.

Reply:
> Do you think that only Int32 needs name distinguished? At present, as there is only int64 support, int64 doesn't say "64" or "int64" in the name. It might be more consistent to name them both with the size.

I think it would make sense to update the name to have them be consistent, e.g. from the [atomic](https://golang.org/pkg/atomic) package.

```go
func AddInt32(addr *int32, delta int32) (new int32)
func AddInt64(addr *int64, delta int64) (new int64)
func AddUint32(addr *uint32, delta uint32) (new uint32)
func AddUint64(addr *uint64, delta uint64) (new uint64)
```

As such, the suggested signatures would become:

```go
func Int32ToFloat(d int32, nBits int) float64
func Int32sToFloats(dst []float64, src []int32, nBits int) []float64
func Int64ToFloat(d int64, nBits int) float64
func Int64sToFloats(dst []float64, src []int64, nBits int) []float64
```

Reply:
While we're at it, may as well add an `IntToFloats`, seeing as `[]int` are used by go-audio: https://godoc.org/github.com/go-audio/audio#IntBuffer.Data

```go
func IntToFloat(d int, nBits int) float64
func IntsToFloats(dst []float64, src []int, nBits int) []float64
```

Reply:
Hehe, this would have been alleviated if Go ever got generics. I don't quite like how [the current Go 2 proposal](https://github.com/golang/proposal/blob/master/design/go2draft-generics-overview.md) looks like though, so it's good that it will be iterated upon.

Reply:
Looks nice that way.  You wanna do it or want me to?

Yes the generics needs to be iterated.   I hope they unify contracts and interfaces and put type arguments in a way that allows control of their instantiations so they can be used template-like and interface-like.  Too busy with zc to say much more though :)

Reply:
> Looks nice that way. You wanna do it or want me to?

I'll push `Int32ToFloat` and `Int32sToFloats` in a second. As for renaming `ToFloat` to `Int64ToFloat`, etc. We can probably discuss it a bit more, to see if there are more places that should be made consistent. E.g. `ToFixed`?

Reply:
I think it would be fine to do the same thing with ToFixed,  but I regret the loss of simplicity of using int64's as it would house everything else.  I guess that loss is a necessary thing to the extent that people are already using int32s, though.